PropertyGraph::Make option that accepts EntityTypeManagers

scober opened the pull request:
If you are going to pass in arrays of type ids to a Make variant, it
makes sense for it to also accept EntityTypeManagers.

roshandathathri commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -530,43 +513,40 @@ class SubgraphQueryMiner {
+
+    QueryEmbeddingListType list = std::move(concurrent_list);
+
+    if (queryGraph.size() <= 1) {
```
It cannot be less than.

ddn0 replied:
just a stylistic thing, I think, unless you mean more by this? I tend to be strong with my conditions on the off chance someone might move a block a code around without noticing the context sensitivity.

roshandathathri replied:
I would prefer an assertion that it is not 0. and then an if condition that it is equal to 1. But it is fine the way it is.

roshandathathri commented on libquery/src/MatchOp.cpp:
```diff
@@ -154,148 +160,105 @@ void galois::MatchOpBuilder::ConstructVarNameToQueryNodeMap() {
-  // prefix sum of edges for nodes
-  std::vector<size_t> prefix_sum;
+bool galois::MatchOpBuilder::ProcessCompilerGraph() {
+  num_query_nodes_ = MaxID(query_nodes_, query_edges_) + 1;
```
Why is number of query nodes dependent on number of query edges? @l-hoang

l-hoang replied:
don't have the full context, but each edge creates at most 2 node and at least 1 if one of the endpoints is an already existing node

unless you were asking something else

ddn0 replied:
I naturally assumed as you @roshandathathri but at some point upstream a query like `a->b` is turned into 1 query edge and 0 query nodes.

l-hoang replied:
oh right

the query node count only counts *singular* nodes not connected to anything. @roshandathathri this is how single node queries were supported IIRC

roshandathathri replied:
Hehe, ok.

ddn0 commented on libquery/include/querying/ops/RegularPathOp.h:
```diff
@@ -0,0 +1,35 @@
+                uint32_t dst_id, std::optional<size_t> start_length,
+                std::optional<size_t> end_length)
+      : start_length_(start_length), end_length_(end_length), label_(label),
+        src_id_(src_id), dst_id_(dst_id), edge_direction_(direction) {}
```
Yeah. Since I haven't implemented the RegularPathOp, I assume that there will be some parts of this that will change. I'll take your comments as notes when I go to implement the regular path op itself.

ddn0 commented on libquery/include/galois/graphs/QueryGraph.h:
```diff
@@ -41,10 +43,10 @@ enum QueryEdgeDirection {
 struct DataNode {
   //! Label on node
-  QueryLabel label;
+  QueryLabel label{};
```
QueryLabel is a `uint32_t`. Since QueryLabel is not a class type, `QueryLabel label` would be default initialization, which would leave label with an indeterminate if it were stack allocated:

https://en.cppreference.com/w/cpp/language/default_initialization

`QueryLabel label{}` is value initialization (though, annoyingly, it has the same syntax as aggregate initialization):

https://en.cppreference.com/w/cpp/language/value_initialization

under value initialization, class types will have their default constructors called and non class types will be zero initialized, so it is a uniform way to do the expected thing and create an object in a known state.

There is some <del>annoying C++</del> subtlety in that value initialization looks like aggregate initialization and under aggregate initialization some members without a default constructor may end up being initialized where under value initialization they would not. You can see the note on the value initialization page for details. But if we assume that default constructors are correct, this won't be an issue in practice.

roshandathathri replied:
Thanks :)

bowu commented:
Are path searches like "-[:Knowns*2.5]->" implemented in the introduced op? I thought it would be handled by MatchOp.

ddn0 commented:
> Are path searches like "-[:Knows*2..5]->" implemented in the introduced op? I thought it would be handled by MatchOp.

This PR doesn't add any new functionality. But I was intending that RegularPathOp would handle such variable paths, unless you think it would be easier for the pattern matcher to do it. The current state is that the pattern matcher doesn't handle any query with `*`. Though certain patterns like `[:LABEL*2]` are converted to `[:LABEL]-[:LABEL]` by the frontend though the current code is marked as broken cc @roshandathathri

bowu commented:
> > Are path searches like "-[:Knows*2..5]->" implemented in the introduced op? I thought it would be handled by MatchOp.
> 
> This PR doesn't add any new functionality. But I was intending that RegularPathOp would handle such variable paths, unless you think it would be easier for the pattern matcher to do it. The current state is that the pattern matcher doesn't handle any query with `*`. Though certain patterns like `[:LABEL*2]` are converted to `[:LABEL]-[:LABEL]` in the frontend; however, the current code is marked as needing more work cc @roshandathathri

My understanding is that we need new Ops to perform single- and multiple-source shortest path queries. Regular paths are "embedded" in the query graph and should be performed inside MatchOp. I'm willing to hear what @roshandathathri and @l-hoang think though.

roshandathathri commented:
> Regular paths are "embedded" in the query graph and should be performed inside MatchOp

I'm fine with either approach. In the short run, maybe this is easier to implement (especially as the plan for regular path op is to return embeddings).

Commit:
PropertyGraph::Make option that accepts EntityTypeManagers

If you are going to pass in arrays of type ids to a Make variant, it
makes sense for it to also accept EntityTypeManagers.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -360,11 +360,15 @@ class KATANA_EXPORT PropertyGraph {
       katana::GraphTopology&& topo_to_assign,
-      NUMAArray<EntityTypeID> node_entity_type_id,
-      NUMAArray<EntityTypeID> edge_entity_type_id) noexcept
+      NUMAArray<EntityTypeID>&& node_entity_type_id,
```
Are we following the model that the type array is kept separate from the manager? Why can't it be a part of the manager?
